src/charm.py

```diff
@@ -63,29 +65,58 @@ def _on_legacy_relation_created(self, event: ops.RelationCreatedEvent) -> None:
         """
         if not self.model.unit.is_leader():
             return
-        self._update_smtp_legacy_relation(event.relation)
+        self._update_relation(event.relation)
 
     def _on_config_changed(self, _) -> None:
         """Handle changes in configuration."""
         self.unit.status = ops.MaintenanceStatus("Configuring charm")
-        self._store_password_as_secret()
+        if self._has_secrets():
+            self._store_password_as_secret()
         self._update_relations()
         self.unit.status = ops.ActiveStatus()
 
-    def _store_password_as_secret(self) -> None:
-        """Store the SMTP password as a secret."""
-        if self._charm_state.password and self._has_secrets():
-            # Note https://github.com/juju/python-libjuju/issues/970
+    def _store_password_as_secret(self) -> Optional[ops.Secret]:
+        """Store the SMTP password as a secret.
+
+        Returns:
+            the secret id.
+        """
+        peer_relation = self.model.get_relation("smtp-peers")
+        assert peer_relation  # nosec
+        if self._charm_state.password and not peer_relation.data[self.app].get("secret-id"):
             secret = self.app.add_secret({"password": self._charm_state.password})
-            self._charm_state.password_id = secret.id
+            peer_relation.data[self.app].update({"secret-id": secret.id})
+            return secret
+        if self._charm_state.password:
+            secret_id = peer_relation.data[self.app].get("secret-id")
+            secret = self.model.get_secret(id=secret_id)
+            secret.set_content({"password": self._charm_state.password})
+            return secret
+        if peer_relation.data[self.app].get("secret-id"):
+            secret_id = peer_relation.data[self.app].get("secret-id")
+            secret = self.model.get_secret(id=secret_id)
+            secret.remove_all_revisions()
+            peer_relation.data[self.app].pop("secret-id")
+        return None
 
     def _update_relations(self) -> None:
         """Update all SMTP data for the existing relations."""
         if not self.model.unit.is_leader():
             return
         for relation in self.model.relations[self.smtp.relation_name]:
-            self._update_smtp_relation(relation)
+            self._update_relation(relation)
         for relation in self.model.relations[self.smtp_legacy.relation_name]:
+            self._update_relation(relation)
+
+    def _update_relation(self, relation: ops.Relation) -> None:
+        """Update the SMTP data for a relation.
+
+        Args:
+            relation: the relation to update the SMTP for.
+        """
+        if relation.name == self.smtp.relation_name:
+            self._update_smtp_relation(relation)
```

Comment:
Why not do the update_smtp_relation and update_smtp_legacy_relation in the for loop?